src/core/iop/memcard.cpp

```diff
@@ -1,6 +1,308 @@
 #include "memcard.hpp"
+#include <cstdint>
+#include <string>
+#include <array>
+#include <vector>
+#include <fstream>
+#include <queue>
 
 Memcard::Memcard()
 {
+	
+}
+
+bool Memcard::is_formatted()
+{
+	return formatted;
+}
+
+bool Memcard::address_in_bounds(uint32_t address, uint32_t length)
+{
+	return address + length <= oem_memcard_bytes;
+}
+
+// Create a file on disk with all bits initialized to their default state, 1.
+// The file is unusable in this state; it must be formatted by a game or BIOS.
+bool Memcard::new_memcard_file(std::string file_name)
+{
+	std::ofstream file(file_name);
+
+	if (!file.is_open())
+	{
+		printf("[MCD] (WARN) Attempted to open ofstream on empty file %s, but it either cannot be opened or could not be created! Aborting...", file_name);
+		return false;
+	}
+
+	// Write 0xff to entire card
+	for (int i = 0; i < oem_memcard_bytes; i++) {
+		file.put(0xff);
+	}
+
+	file.close();
+
+	return true;
+}
+
+// Open the memcard file_name, mounting its contents to this memcard.
+bool Memcard::open(std::string file_name)
+{
+	std::ifstream file(file_name);
+
+	if (!file.is_open())
+	{
+		printf("[MCD] (WARN) Attempted to open ifstream on memcard file %s, but it either cannot be opened or does not exist! Aborting...", file_name);
+		return false;
+	}	
+
+	file.read(raw_data.data(), oem_memcard_bytes);
+
+	if (file)
+		printf("[MCD] (INFO) Successfully read all bytes into memory!");
+	else
+		printf("[MCD] (WARN) Failed to read all bytes into memory! Only read %d!", file.gcount());
+
+	file.close();
+	this->file_name = file_name;
+	this->formatted = true;
```

Comment:
This and the next few lines can be replaced with `this->formatted = std::equal(format_string.begin(), format_string.end(), raw_data);`